Classes/common/touchdb/TDPuller.m

```diff
@@ -347,35 +366,41 @@ - (void)pullRemoteRevision:(TD_Revision*)rev
-                                           // OK, now we've got the response revision:
-                                           if (error) {
-                                               strongSelf.error = error;
-                                               [strongSelf revisionFailed];
-                                               strongSelf.changesProcessed++;
-                                           } else {
-                                               TD_Revision* gotRev =
-                                                   [TD_Revision revisionWithProperties:dl.document];
-                                               gotRev.sequence = rev.sequence;
-                                               // Add to batcher ... eventually it will be fed to
-                                               // -insertRevisions:.
-                                               [_downloadsToInsert queueObject:gotRev];
-                                               [strongSelf asyncTaskStarted];
-                                           }
-
-                                           // Note that we've finished this task:
-                                           [strongSelf removeRemoteRequest:dl];
-                                           [strongSelf asyncTasksFinished:1];
-                                           --_httpConnectionCount;
-                                           // Start another task if there are still revisions
-                                           // waiting to be pulled:
-                                           [strongSelf pullRemoteRevisions];
-                                       }];
-    [self addRemoteRequest:dl];
+    dl = [[TDMultipartDownloader alloc] initWithURL: TDAppendToURL(_remote, path)
+                                           database: _db
+                                     requestHeaders: self.requestHeaders
+                                       onCompletion:
+        ^(TDMultipartDownloader* dl, NSError *error) {
+            __strong TDPuller *strongSelf = weakSelf;
+            // OK, now we've got the response revision:
+            if (error) {
+                // if ignoreMissingDocs is true, we know that some requests might 404
+                if (!(ignoreMissingDocs && error.code == 404)) {
+                    strongSelf.error = error;
+                    [strongSelf revisionFailed];
+                }
+                strongSelf.changesProcessed++;
+            } else {
+                TD_Revision* gotRev = [TD_Revision revisionWithProperties: dl.document];
+                gotRev.sequence = rev.sequence;
+                // Add to batcher ... eventually it will be fed to -insertRevisions:.
+                if (immediatelyInsert) {
+                    [strongSelf insertDownloads:@[gotRev]];
+                } else {
+                    [_downloadsToInsert queueObject: gotRev];
+                }
+                [strongSelf asyncTaskStarted];
```

Comment:
Turned out immediatelyInsert decrements the asynctaskcount so we need to match here with an increment.
Though, it occurs to me we decrement asynctaskcount prior to incrementing it?